App review prompt

mikoldin123 opened the pull request:
### Description
PR added a feature to show app review dialog prompts on specific triggers in the app.

- Show survey
- App review rating

mpretty-cyro reviewed (changes requested):
- [ ] Would also be nice to add some unit tests for the `HomeViewModel` to validate the state updates as expected to show this prompt when the various defaults are updated

mpretty-cyro reviewed (changes requested):
- [x] The spacing and font weights don't match the design

**Design:**
<img width="410" height="225" alt="Image" src="https://github.com/user-attachments/assets/84033d96-39c2-48c2-ab45-18e809249abc" />

**Implementation:**
<img width="410" height="221" alt="Image" src="https://github.com/user-attachments/assets/61597520-2b0d-4e34-92e6-0672937abd67" />

mikoldin123 commented on Session/Home/HomeVC.swift:
```diff
@@ -790,3 +832,29 @@ public final class HomeVC: BaseVC, LibSessionRespondingViewController, UITableVi
+    
+    func onHandleSecondayTappedForState(_ state: AppReviewPromptState) {
+        switch state {
+        case .feedback, .rateSession: viewModel.handlePromptChangeState(nil)
```
Thanks, will set this config

mikoldin123 commented on Session/Home/HomeViewModel.swift:
```diff
@@ -76,6 +83,8 @@ public class HomeViewModel: NavigatableStateHolder {
         let loadedPageInfo: PagedData.LoadedInfo<SessionThreadViewModel.ID>
         let itemCache: [String: SessionThreadViewModel]
+        let appReviewPromptState: AppReviewPromptState?
+        let appReviewPromptTimestamp: TimeInterval?
```
`appReviewPromptTimestamp ` is currently needed, it's checked along with the prompt state to handle the on appear behavior in presenting the prompts

mpretty-cyro commented on SessionUIKit/Style Guide/Constants.swift:
```diff
@@ -15,4 +15,6 @@ public enum Constants {
     public static let session_network_data_price: String = "Price data powered by CoinGecko<br/>Accurate at {date_time}"
     public static let app_pro: String = "Session Pro"
+    public static let store_variant: String = "App Store"
+    public static let feedback_url: String = "https://www.surveymonkey.com/r/YLDZJR8"
```
The `store_variant` value here will have the same issue

mikoldin123 commented on Session/Home/HomeViewModel.swift:
```diff
@@ -196,6 +213,17 @@ public class HomeViewModel: NavigatableStateHolder {
+            if retryCount == 0, let retryDate = dependencies[defaults: .standard, key: .rateAppRetryDate], dependencies.dateNow >= retryDate {
+                dependencies[defaults: .standard, key: .rateAppRetryDate] = nil
+                dependencies[defaults: .standard, key: .rateAppRetryAttemptCount] = 1
+                dependencies[defaults: .standard, key: .didShowAppReviewPrompt] = false
```
Got it will make the changes. Thanks @mpretty-cyro

mpretty-cyro reviewed (changes requested):
- [x] There still seems to be an issue with the padding in the UI - it seems that something weird is going on with the scaling in the Figma doc but this component should match the styling from `ConfirmationModal` so please use the same fonts and padding as in there:
**Title:** `.boldSystemFont(ofSize: Values.mediumFontSize)`
**Body:** `ConfirmationModal.explanationFont`
**Button:** `.boldSystemFont(ofSize: Values.mediumFontSize)` (or just use `Modal.createButton` to create these

The `contentStack` should use the following:
```Swift
result.axis = .vertical
result.spacing = Values.smallSpacing
result.isLayoutMarginsRelativeArrangement = true
result.layoutMargins = UIEdgeInsets(
    top: Values.largeSpacing,
    left: Values.veryLargeSpacing,
    bottom: Values.verySmallSpacing,
    right: Values.veryLargeSpacing
)
```
And you should just `contentStack.pin(to: self)` (no additional padding)

Then the close button should be positioned with:
```
closeButton.pin(.top, to: .top, of: self, withInset: Values.smallSpacing)
closeButton.pin(.right, to: .right, of: self, withInset: -Values.smallSpacing)
```
And in the confirmation modal it's defined with:
```
private lazy var closeButton: UIButton = UIButton(primaryAction: UIAction { [weak self] _ in self?.close() })
    .withConfiguration(
        UIButton.Configuration
            .plain()
            .withImage(UIImage(named: "X")?.withRenderingMode(.alwaysTemplate))
            .withContentInsets(NSDirectionalEdgeInsets(top: 6, leading: 6, bottom: 6, trailing: 6))
    )
    .withConfigurationUpdateHandler { button in
        switch button.state {
            case .highlighted: button.imageView?.tintAdjustmentMode = .dimmed
            default: button.imageView?.tintAdjustmentMode = .normal
        }
    }
    .withImageViewContentMode(.scaleAspectFit)
    .withThemeTintColor(.textPrimary)
    .withAccessibility(
        identifier: "Close button",
        label: "Close button"
    )
    .with(.width, of: ConfirmationModal.closeSize)
    .with(.height, of: ConfirmationModal.closeSize)
```

mikoldin123 commented:
Added the following updates:

- Show review prompt when force closing app
- Added 24hr buffer on the review retry date check
- Added check if app was installed or updated to app review version
- Fix content and button colors to pass standards for ocean light theme
- Updated displayed survey url to full path

mpretty-cyro commented on Session/Home/HomeViewModel.swift:
```diff
@@ -52,18 +53,48 @@ public class HomeViewModel: NavigatableStateHolder {
+        let buffer: TimeInterval = 24 * 60 * 60
 
-        if retryCount == 0, let retryDate = dependencies[defaults: .standard, key: .rateAppRetryDate], dependencies.dateNow >= retryDate {
+        if retryCount == 0, let retryDate = dependencies[defaults: .standard, key: .rateAppRetryDate], dependencies.dateNow.timeIntervalSince(retryDate) >= -buffer {
```
This logic is now getting fairly complex for `HomeViewModel.init` - can you please add a static function: `AppReviewPromptModel.loadInitialState` which returns a tuple of `AppReviewPromptState?` and put this logic in there?

mpretty-cyro commented on Session/Home/HomeViewModel.swift:
```diff
@@ -52,18 +53,48 @@ public class HomeViewModel: NavigatableStateHolder {
         }
         
+        // Checks if new version of app is from install or update
+        var didInstallAppReviewPromptVersion: Bool {
```
I think `appWasInstalledPriorToAppReviewRelease` would be a more self-explanatory name for this variable (you'll have to invert the logic  though)

mpretty-cyro replied:
Can you please make this a `static func` on the `AppReviewPromptModal` similar to the `loadInitialState` above (means we can clean it out of `HomeViewModel.init` as well)

mpretty-cyro commented on SessionUIKit/Style Guide/Constants+Apple.swift:
```diff
@@ -11,4 +11,8 @@ public extension Constants {
+    static let platform_name = "iOS"
+    
+    // MARK: -
+    static let review_prompt_availability_version = "2.14.1"
```
Can you please move this to be a static variable in `AppReviewPromptModel`? (It's unlikely that it would ever be used elsewhere so it can live in there)

Additionally we probably need to add a `TODO` or a task to ensure we update this value prior to the release (as `2.14.1` is already released, and `2.14.2` is probably going to to out early next week which might be too soon for this PR to get merged - will probably be `2.14.3` pending QA)

mpretty-cyro commented on Session/Home/HomeViewModel.swift:
```diff
@@ -578,16 +617,44 @@ public class HomeViewModel: NavigatableStateHolder {
+            )
+            
+            if !dependencies[feature: .simulateAppReviewLimit] {
+                SKStoreReviewController.requestReview(in: scene)
```
So it seems that `SKStoreReviewController` is deprecated (https://developer.apple.com/documentation/storekit/skstorereviewcontroller) - can you please update this to use `RequestReviewAction` instead? (https://developer.apple.com/documentation/storekit/requestreviewaction)

mpretty-cyro commented on Session/Home/HomeViewModel.swift:
```diff
@@ -578,16 +617,44 @@ public class HomeViewModel: NavigatableStateHolder {
     }
     
+    @objc
+    func systemHasPresentedNewWindow(notification: Notification) {
```
Can you please rename this to `windowBecameVisibleAfterTriggeringAppStoreReview` (the current name is very generic but the functionality is tied to the review process so the name should probably be specific)

mpretty-cyro commented on Session/Home/HomeViewModel.swift:
```diff
@@ -543,6 +578,7 @@ public class HomeViewModel: NavigatableStateHolder {
         DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(1)) { [self, dependencies] in
             // Set flag that review prompt was already presented
             dependencies[defaults: .standard, key: .didShowAppReviewPrompt] = true
+            dependencies[defaults: .standard, key: .didIgnoreAppReviewPrompt] = true
```
So I feel like this flag is a little backwards (ie. setting `didIgnoreAppReviewPrompt` when the prompt appears isn't entirely accurate as the user hasn't had the chance to actually see the prompt yet)

I wonder whether instead we should have a `didActionAppReviewPrompt` which gets set to `true` when the users presses one of the buttons (should it also be set when pressing the `X`?) and set to `false` prior to displaying the prompt - that way if we tie any future logic to this flag it is being set at more intuitive times

mpretty-cyro commented on Session/Home/HomeViewModel.swift:
```diff
@@ -578,16 +617,44 @@ public class HomeViewModel: NavigatableStateHolder {
+
+            // Added 2 sec delay to give time for requet review to proc
+            DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(2)) { [self] in
+                NotificationCenter.default.removeObserver(self, name: UIWindow.didBecomeVisibleNotification, object: nil)
```
While it doesn't currently happen, we should probably add `NotificationCenter.default.removeObserver(self)` to the `HomeViewModel.deinit` just in case it gets destroyed during this delay

mpretty-cyro commented on Session/Home/HomeViewModel.swift:
```diff
@@ -578,16 +617,44 @@ public class HomeViewModel: NavigatableStateHolder {
+            }
+
+            // Added 2 sec delay to give time for requet review to proc
+            DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(2)) { [self] in
```
Please change this to a `weak self` capture - it's not an issue with the current code but capturing `self` strongly means if there were future changes to the code it'd be easier to miss a retain cycle (you can either `guard let self = self else { return }` of `NotificationCenter.default.removeObserver(self as Any, ...` to resolve the warning that this creates)

mpretty-cyro commented on Session/Home/App Review/View/AppReviewPromptDialog.swift:
```diff
@@ -149,6 +149,15 @@ class AppReviewPromptDialog: UIView {
+        
+        let isButtonsHidden = primaryButton.isHidden && secondaryButton.isHidden
+        
+        buttonStack.layoutMargins = .init(
```
Can you please remove the `result.layoutMargins = ...` from the `lazy var buttonStack` creation (since it's always going to be overwritten here having there is redundant)

mpretty-cyro commented on Session/Home/App Review/AppReviewPromptModel.swift:
```diff
@@ -22,30 +22,38 @@ extension AppReviewPromptModel {
+        let retryCount = dependencies[defaults: .standard, key: .rateAppRetryAttemptCount]
 
+        // A buffer of 1 hour
+        let buffer: TimeInterval = 60 * 60
```
There is a comment below this that says `This block will execute if the current time is within 24 hours of the retryDate` which will need to be updated to match this new buffer period

mpretty-cyro commented on Session/Home/App Review/AppReviewPromptModel.swift:
```diff
@@ -22,30 +22,38 @@ extension AppReviewPromptModel {
+        return promptState
+    }
+    
+    static func shouldShowAppReviewAgain(using dependencies: Dependencies) -> Bool {
```
This name is misleading - it suggests that it's going to return a flag indicating whether the prompt should be shown but instead it's updating a bunch of values in `userDefaults`, maybe rename it to something like `checkAndRefreshAppReviewState`
